python/ray/ml/train/integrations/huggingface/huggingface_trainer.py

```diff
@@ -146,52 +174,391 @@ def trainer_init_per_worker(train_dataset, eval_dataset, **config):
+        trainer_init_per_worker: Callable[
+            [TorchDataset, Optional[TorchDataset], Any], transformers.trainer.Trainer
+        ],
+    ):
         def train_loop_per_worker(config):
-            # Set to None just to make CI pass & show
-            # the intended usage with trainer_init_per_worker
-            train_dataset = None
-            eval_dataset = None
-            trainer = trainer_init_per_worker(train_dataset, eval_dataset, **config)
-            trainer.train()
+            # Env vars necessary for HF to setup DDP
+            os.environ["RANK"] = str(train.world_rank())
+            os.environ["WORLD_SIZE"] = str(train.world_size())
+            os.environ["LOCAL_RANK"] = str(train.local_rank())
+
+            train_dataset = train.get_dataset_shard(TRAIN_DATASET_KEY)
+            eval_dataset = train.get_dataset_shard(EVALUATION_DATASET_KEY)
+            train_columns = set(train_dataset.schema(fetch_if_missing=True).names)
+            if "labels" not in train_columns:
+                raise ValueError(
+                    "'labels' column must be present in the training dataset!"
+                )
+            train_columns.remove("labels")
+            if eval_dataset:
+                eval_columns = set(eval_dataset.schema(fetch_if_missing=True).names)
+                if "labels" not in eval_columns:
+                    raise ValueError(
+                        "'labels' column must be present in the evaluation dataset!"
+                    )
+                eval_columns.remove("labels")
+
+                if not eval_columns.issuperset(train_columns):
+                    raise ValueError(
+                        "Evaluation dataset must have a superset of the columns in "
+                        "the training dataset. "
+                        f"Missing columns: {list(train_columns - eval_columns)}"
+                    )
+
+            # HF-supported format
+            feature_columns = {column: [column] for column in train_columns}
+
+            # we use batch size 1 here, as it will be converted to
+            # desired size inside transformers.Trainer. Possible optimization
+            # in the future
+            batch_size = 1
+            train_torch_dataset = _process_dataset_for_hf(
+                train_dataset, feature_columns, batch_size=batch_size
+            )
+
+            if eval_dataset:
+                eval_torch_dataset = _process_dataset_for_hf(
+                    eval_dataset, feature_columns, batch_size=batch_size
+                )
+            else:
+                eval_torch_dataset = None
+
+            # ensure no HF logging callbacks are added
+            # aside from doubling functionality with our callbacks,
+            # the Wandb callbacks causes training to freeze
+            # TODO(yard1): Automatically set `no_cuda`
+            with patch(
+                "transformers.trainer.get_reporting_integration_callbacks", lambda x: []
+            ):
```

Comment:
Monkey patching is not great, can we instead do `trainer.pop_callback()` after initialization for all non-default callbacks?

Reply:
Done